### PR TITLE
docs: Document Merge Tokenizer Pipeline

### LINKED ISSUE #296: Model Documentation : Merge tokenizer

### PR DESCRIPTION
**Pull request by MoosaTae**
## Why this PR
Why we need this PR?

## Changes
- Write some changes here

## Related Issues
Close #296 

## Checklist
- [ ] PR should be in the [Naming convention](../../docs/PR_NAMING.md)
- [ ] Assign yourself in to Assigneees
- [ ] Tag related issues
- [ ] Constants name should be ALL_CAPITAL, function name should be snake_case, and class name should be CamelCase
- [ ] complex function/algorithm should have [Docstring](https://peps.python.org/pep-0257/)
- [ ] 1 PR should not have more than 200 lines changes (Exception for test files). If more than that please open multiple PRs
- [ ] At least PR reviewer must come from the task's team (model, eval, data)

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -11,4 +11,5 @@ src/data/**/full_input*
 /pythainlp-data
 /outputs
 /etc/
-/image_sandbox
\ No newline at end of file
+/image_sandbox
+.DS_Store
```
Comment (new5558) [inline .gitignore:15]:
```suggestion
**/.DS_Store
```

File: src/model/scripts/llama_thai_tokenizer/README.md [hunk 2]
```diff
@@ -1,25 +1,47 @@
-# Tokenizer Experiment
+# OpenThaiGPT Merged Tokenizer Pipeline
 
-merge LLaMa Tokenizer with [wangchanberta](https://huggingface.co/airesearch/wangchanberta-base-att-spm-uncased/tree/main) Tokenizer by extent vocabulary following [working example](https://github.com/ymcui/Chinese-LLaMA-Alpaca/blob/main/README_EN.md) 
+we plan to use [Pretrained LLaMA Model](https://ai.meta.com/blog/large-language-model-llama-meta-ai/) as a base model for finetuning but LLaMa Tokenizer (BPE Tokenizer) have less Thai vocabulary, this pipeline intent to extent vocabulary of LlaMA tokenizer, using merge method following [Chinese-LLaMA-Alpaca](https://github.com/ymcui/Chinese-LLaMA-Alpaca/blob/main/README_EN.md)
```
Comment (new5558) [inline src/model/scripts/llama_thai_tokenizer/README.md:3]: `this pipeline intent to` -> `This pipeline intend to`

Comment (new5558) [inline src/model/scripts/llama_thai_tokenizer/README.md:3]: `we plan` -> `We plan`

File: src/model/scripts/llama_thai_tokenizer/README.md [hunk 3]
```diff
@@ -1,25 +1,47 @@
...
-### Results
-- Text : การใช้งานหลักของ LLaMA คือการวิจัยเกี่ยวกับรูปแบบภาษาที่ใหญ่
-    - Tokenized by LLaMA tokenizer: ['▁', 'ก', 'า', 'ร', '<0xE0>', '<0xB9>', '<0x83>', 'ช', '้', 'ง', 'า', 'น', 'ห', 'ล', 'ั', 'ก', 'ข', 'อ', 'ง', '▁L', 'La', 'MA', '▁', 'ค', 'ื', 'อ', 'ก', 'า', 'ร', 'ว', 'ิ', 'จ', 'ั', 'ย', 'เ', 'ก', 'ี', '่', 'ย', 'ว', 'ก', 'ั', 'บ', 'ร', 'ู', 'ป', 'แ', 'บ', 'บ', 'ภ', 'า', 'ษ', 'า', 'ท', 'ี', '่', '<0xE0>', '<0xB9>', '<0x83>', 'ห', 'ญ', '่']
-    - Tokenized by English-Thai LLaMA tokenizer: ['▁การใช้งาน', 'หลักของ', '▁L', 'La', 'MA', '▁คือการ', 'วิจัย', 'เกี่ยวกับ', 'รูปแบบ', 'ภาษา', 'ที่ใหญ่']
-    - EngOnly time: 0.00027060508728027344
-    - EngThai time: 0.00016260147094726562
-    - time diff: 0.00010800361633300781
+### Results after merging
 
+- Text : การใช้งานหลักของ LLaMA คือการวิจัยเกี่ยวกับรูปแบบภาษาที่ใหญ่
```
Comment (new5558) [inline src/model/scripts/llama_thai_tokenizer/README.md:42]: I can only see Thai result. Where is English result?

File: src/model/scripts/llama_thai_tokenizer/README.md [hunk 4]
```diff
@@ -1,25 +1,47 @@
...
+
+argument
+
+- model_name: name of LLaMA model that you use
+- output_path: path to save
+
+2. merge tokenizer `need to have Thai BPE tokenizer first` then add path of it be an argument of following script
+
+```bash
+python3 scripts/llama_thai_tokenizer/merge_tokenizer.py --llama_path path/to/model --thai_sp_path path/to/thai_model --output_path path/to/save
```
Comment (new5558) [inline src/model/scripts/llama_thai_tokenizer/README.md:26]:
```suggestion
python merge_tokenizer.py --llama_path <llama_model_path> --thai_sp_path <spm_model_path> --output_path <output_tokenizer_path>
```

File: src/model/scripts/llama_thai_tokenizer/README.md [hunk 5]
```diff
@@ -1,25 +1,47 @@
...
+```
+
+argument
+
+- llama_path: path to LLaMA tokenizer
```
Comment (new5558) [inline src/model/scripts/llama_thai_tokenizer/README.md:31]: When you mentioned path, does this mean `model.spm` or `huggingface_tokenizer_path` (folder)?

Can you elaborate on it here in this argument for each paths?

File: src/model/scripts/llama_thai_tokenizer/README.md [hunk 6]
```diff
@@ -1,25 +1,47 @@
...
+- thai_sp_path: path to Thai BPE tokenizer
+- output_path: path to save
 
 ## To test merged tokenizer
 
 1.  run llama_thai_token_test.py and inference time checked
-```python3 scripts/llama_thai_tokenizer/time_inference_check.py```
+    `python3 scripts/llama_thai_tokenizer/time_inference_check.py`
```
Comment (new5558) [inline src/model/scripts/llama_thai_tokenizer/README.md:38]:
```suggestion
    `python time_inference_check.py`
```

File: src/model/scripts/llama_thai_tokenizer/README.md [hunk 7]
```diff
@@ -1,25 +1,47 @@
...
+- merge LLaMA tokenizer by extenting vocabulary and merge rule from BPE Thai tokenizer
+  ![method](merge_method.png)
 
-```python3 scripts/llama_thai_tokenizer/merge_tokenizer.py```
+## merge and save merge tokenizer
+
+1. load LLaMA tokenizer by pass your model name to an argument
+
+```bash
+python3 src/model/scripts/llama_thai_tokenizer/load_tokenizer.py --model_name meta-llama/Llama-2-7b --output_path path/to/save
```
Comment (new5558) [inline src/model/scripts/llama_thai_tokenizer/README.md:15]:
```suggestion
python load_tokenizer.py --model_name meta-llama/Llama-2-7b --output_path <save_path>
```

File: .gitignore (hunk 1 above)
Comment (MoosaTae) [reply .gitignore]: done

File: src/model/scripts/llama_thai_tokenizer/README.md (hunk 2 above)
Comment (MoosaTae) [reply src/model/scripts/llama_thai_tokenizer/README.md]: done

Comment (MoosaTae) [reply src/model/scripts/llama_thai_tokenizer/README.md]: done

File: src/model/scripts/llama_thai_tokenizer/README.md (hunk 3 above)
Comment (MoosaTae) [reply src/model/scripts/llama_thai_tokenizer/README.md]: done

File: src/model/scripts/llama_thai_tokenizer/README.md (hunk 7 above)
Comment (MoosaTae) [reply src/model/scripts/llama_thai_tokenizer/README.md]: done

File: src/model/scripts/llama_thai_tokenizer/README.md (hunk 4 above)
Comment (MoosaTae) [reply src/model/scripts/llama_thai_tokenizer/README.md]: done

File: src/model/scripts/llama_thai_tokenizer/README.md (hunk 6 above)
Comment (MoosaTae) [reply src/model/scripts/llama_thai_tokenizer/README.md]: done

File: src/model/scripts/llama_thai_tokenizer/README.md (hunk 5 above)
Comment (MoosaTae) [reply src/model/scripts/llama_thai_tokenizer/README.md]: done

**Comment by new5558** [comment]
Can you temporarily pause pytest by adding @fixture on a test case that has problem
https://docs.pytest.org/en/6.2.x/fixture.html
image

**Review by boss-chanon — APPROVED** [review]

**Review by kwankoravich — APPROVED** [review]

File: src/model/scripts/llama_thai_tokenizer/README.md [hunk 8]
```diff
@@ -1,25 +1,57 @@
-# Tokenizer Experiment
+# OpenThaiGPT Merged Tokenizer Pipeline
 
-merge LLaMa Tokenizer with [wangchanberta](https://huggingface.co/airesearch/wangchanberta-base-att-spm-uncased/tree/main) Tokenizer by extent vocabulary following [working example](https://github.com/ymcui/Chinese-LLaMA-Alpaca/blob/main/README_EN.md) 
+We plan to use [Pretrained LLaMA Model](https://ai.meta.com/blog/large-language-model-llama-meta-ai/) as a base model for finetuning but LLaMa Tokenizer (BPE Tokenizer) have less Thai vocabulary, this pipeline intend to extent vocabulary of LlaMA tokenizer, using merge method following [Chinese-LLaMA-Alpaca](https://github.com/ymcui/Chinese-LLaMA-Alpaca/blob/main/README_EN.md)
```
[the remark this answers is not in the document]
Comment (MoosaTae) [reply src/model/scripts/llama_thai_tokenizer/README.md]: done

**Pull request merged**